pkg/dmesg/watcher.go

```diff
@@ -0,0 +1,197 @@
+package dmesg
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"regexp"
+	"strings"
+	"time"
+
+	"github.com/leptonai/gpud/log"
+	"github.com/leptonai/gpud/pkg/process"
+)
+
+var DefaultWatchCommands = []string{
+	"dmesg --decode --time-format=iso --nopager --buffer-size 163920 -w || true",
+
+	// run last commands as fallback, in case "dmesg -w" flag only works in some machines
+	"dmesg --decode --time-format=iso --nopager --buffer-size 163920 -W || true",
+}
+
+type LogLine struct {
+	Timestamp time.Time
+	Facility  string
+	Level     string
+	Content   string
+
+	// Error is set when a dmesg command fails.
+	Error string
+}
+
+type Watcher interface {
+	// Watch returns a channel that emits log lines.
+	// The channel is closed on (1) process exit, (2) on calling "Close" method
+	Watch() <-chan LogLine
+	// Closes the existing/ongoing watch/stream routines.
+	// Safe to call multiple times.
+	Close()
+}
+
+func NewWatcher() (Watcher, error) {
+	return newWatcher(DefaultWatchCommands...)
+}
+
+func newWatcher(cmds ...string) (Watcher, error) {
+	if len(cmds) == 0 {
+		return nil, errors.New("no commands provided")
+	}
+
+	ctx, cancel := context.WithCancel(context.Background())
+	ch, err := watch(ctx, cmds...)
+	if err != nil {
+		cancel()
+		return nil, err
+	}
+	return &watcher{ch: ch, cancel: cancel}, nil
+}
+
+type watcher struct {
+	ch     <-chan LogLine
+	cancel context.CancelFunc
+}
+
+func (w *watcher) Watch() <-chan LogLine {
+	return w.ch
+}
+
+func (w *watcher) Close() {
+	w.cancel()
+}
+
+func watch(
+	ctx context.Context,
+	cmds ...string,
+) (<-chan LogLine, error) {
+	ch := make(chan LogLine, 1000)
+	p, err := process.New(
+		process.WithCommand(cmds...),
```

Comment:
do we simply run the `dmesg -w` and `dmesg -W` commands? I thought we should try `-w` first, and if it fails, we merge the outputs of `dmesg` and `dmesg -W`

Reply:
Yes

```
dmesg --decode --time-format=iso --nopager --buffer-size 163920 -w || true
dmesg --decode --time-format=iso --nopager --buffer-size 163920 -W
```

so whatever works it blocks

Reply:
i mean we should merge the outputs of `dmesg` and `dmesg -W` if the first line fails

Reply:
@ccding Yes, already doing that. 

```go
var DefaultWatchCommands = []string{
	"dmesg --decode --time-format=iso --nopager --buffer-size 163920 -w || true",

	// run last commands as fallback, in case "dmesg -w" flag only works in some machines
	"dmesg --decode --time-format=iso --nopager --buffer-size 163920 -W || true",
}

func NewWatcher() (Watcher, error) {
	return newWatcher(DefaultWatchCommands...)
}
```

`TestWatchMultipleCommands` is testing this behavior :)